mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala

```diff
@@ -538,7 +538,7 @@ private[ml] object GBTClassifierParams {
     Array("logistic").map(_.toLowerCase(Locale.ROOT))
 }
 
-private[ml] trait GBTClassifierParams extends GBTParams with TreeClassifierParams {
+private[ml] trait GBTClassifierParams extends GBTParams with TreeRegressorParams {
```

Comment:
Why does it need regressor params to fix this? it sounds like it should extend TreeClassifierParams, just given the name

Reply:
Yes, by the names seems so, but if you check their implementation, they both just add a `impurity` param with the only difference of the allowed values. In the case of `TreeRegressorParams` the only allowed value is `variance` (which is what `GBTClassifier` uses), in `TreeClassifierParams` the allowed values are `gini` (default) and `entropy`. So `variance` is not a valid value for the `impurity` param of `TreeClassifierParams`, so there is no way to set it properly for the `GBTClassifier`....

Reply:
Surely that's not the ideal way to fix it though, semantically. Can't GBTClassifier override the impurity internally (or can we let it do so)?

Reply:
To continue to extend `TreeClassifierParams` but be able to use correct `impurity`, seems we need to remove `final` from `impurity` in trait `TreeClassifierParams`...

Reply:
I will try a small refactor creating a `HasImpurity` trait with the allowed values being overridden by its descendant. This way it should work. Let me know what you think about it. Thanks.

Reply:
@srowen I think the only alternative here us to avoid extending either of them and define the `impurity` param directly into this trait. What do you think?

Reply:
That's preferable to extending the 'wrong' trait. Hm, can TreeClassifierParams not just allow all impurity values? and have GBTClassifierParams set it to "variance" in the constructor? that sounds more correct. I know you guys have looked into it more but I still wasn't clear why this wasn't an option.

Reply:
@srowen the problem is the `setDefault` which is performed in the `TreeClassifierParams` if I do what you are suggesting, then you get an exception because `gini` is not a valid value for it. Unless we put as allowed values `gini` and `variance` and we set it back to `variance` in the GBTClassifier but that is way too hacky IMHO. I'll go ahead with adding the `impurity` directly in `GBTClassifierParams` then, thanks.

Reply:
Wait, why can't all three values be valid? and set the default to gini in the parent, and set the actual value to variance in the base class? I am still not sure why this isn't possible and that on the contrary sounds like the most natural solution.

Reply:
Because in GBTClassifier the only valid value is `variance`. It is true that the `setImpurity` method throws an exception if someone tries to set it, but we are not safe against bad values as it happens now. And, most importantly `TreeClassifierParams` is extended by other classes, so we would allow the `variance` value to be set for RandomForest and DecisionTree, which is bad because it is not a valid value for them.

Reply:
Yes, other classes have to restrict what's settable. I don't see the issue with bad values here? Is it serialized forms? At least this sounds like the normal OOP way to solve this. Other ways around this sound like hacks.

Reply:
I mean restricting in the `setImpurity` method as you are suggesting is probably feasible, but I am not sure it is the right thing to do. I haven't seen that pattern anywhere else in the codebase. We set the allowed parameters in the Param itself usually. One of the reason is to avoid problems when you are setting the default for instance or when `set` is used (this happens for instance when loading a saved model, so we can load an invalid model): so I think the solution is not really safe.

I am not sure the solution you are proposing is cleaner. Also, before this PR there were 2 interfaces: `TreeClassifierParams` and `TreeRegressorParams` only for this reason, ie. for the impurity param to allow/have different values. We could switch to a single `HasImpurity` with your approach.

Reply:
I looked more into this and see that setImpurity is deprecated. I presume the point is to use set(impurity, ...) instead. Yeah, that's no longer possible to override in subclasses, scratch that; overriding setImpurity would have been just fine IMHO but that's not going to last anyway.

I do agree this is why there are different traits for classifiers and regressors, but I don't think that means a classifier should extend TreeRegressorParams because its parameters happen to match.

One option is to let the definition of `impurity` itself be overridden. That seems OK. Or we could make a new 'VarianceClassifier' or something that defines this variance-only impurity parameter and let TreeRegressionParams and GBTClassifierParams extend it. It is a little wacky, but quite reasonable from an OOP perspective. I think.

Reply:
@srowen yes, I'll try and create a `VarianceClassifier` trait. Thanks.